sympy/tensor/array/arrayop.py

```diff
@@ -208,3 +208,41 @@ def derive_by_array(expr, dx):
             return ImmutableDenseNDimArray([expr.diff(i) for i in dx], dx.shape)
         else:
             return diff(expr, dx)
+
+
+def _array_transpose(expr, order=None):
+    """
+    Evaluates the transpose of an array.
+
+    Optional parameter to specify an index permutation.
+    """
+    if not isinstance(expr, NDimArray):
+        return transpose(expr)
+
+    rank = expr.rank()
+
+    if order is None:
+        # order is reversed indices order:
+        order = tuple(reversed(range(expr.rank())))
+    elif len(order) < rank:
+        # order is incomplete, fill it:
+        order = tuple(order) + tuple(range(len(order), rank))
+
```

Comment:
I'd suggest to raise an exception if
`tuple(sorted(order)) != tuple(range(expr.rank))`